include/gatekeeper_fib.h

```diff
@@ -215,7 +220,12 @@ int del_fib_entry_numerical(
 	struct ip_prefix *prefix_info, struct gk_config *gk_conf);
 int del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf);
 
-/* TODO Customize the hash function for IPv4. */
+static inline uint32_t
+custom_ipv4_hash_func(const void *key,
+	__attribute__((unused)) uint32_t length, uint32_t initval)
+{
+        return *(const uint32_t *)key % initval;
```

Comment:
1. How did you evaluate this hash function? Isn't going to generate lots of collisions?

2. One doesn't need to mod the result, the hash table will already do it with any hash.

3. The function is being passed as a parameter of a function, so the compiler is ignoring the tag `inline` and duplicating binary code. Move the code the the .c file and export it to avoid duplicate binary code.

Reply:
1. In our current implementation, the number of entries is equal to the number of slots in the hash table, so there should be no collision.

2. You are right. The hash table will do that, so the mod can be removed.

Reply:
1. Your argument is not sound because the number of buckets, the keys, AND the hash function establish the collisions. For example, consider a LAN network with the prefix 10.0.0.0/24. The IP addresses only change in the last octet, so depending which bits are the most important for the implementation of the hash table (i.e. the most or less significant bits), we'll have more or less collision with the hash function you're proposing. Notice that this would even change with the machine being little or big endian! We need a better custom hash function.

Reply:
Thanks for the explanation! You are right, this would change with the machine being little or big endian. For the IP prefixes, shouldn't the hash function always take the less significant bits of the IP address as the most important? Since the hash table key used is in network byte order, should we translate the key to be host byte order?

Reply:
> For the IP prefixes, shouldn't the hash function always take the less significant bits of the IP address as the most important?

Yes, the least significat bits of a given IP prefix are the most important for entropy.

If the DPDK's implementation takes a mod over the hash, what I expect it does, the most important bits of a hash function are the least significant bits as well. But the point is that it depends on the implementation of the hash table (in this case DPDK) and matching the byte order of an IP address and of the host.

> Since the hash table key used is in network byte order, should we translate the key to be host byte order?

Yes.

Reply:
> If the DPDK's implementation takes a mod over the hash, what I expect it does, the most important bits of a hash function are the least significant bits as well. But the point is that it depends on the implementation of the hash table (in this case DPDK) and matching the byte order of an IP address and of the host.

That makes sense! After checking the DPDK code, the bucket indexes are obtained using mod (i.e., ``sig & h->bucket_bitmask``, where ``h->bucket_bitmask = h->num_buckets - 1``, and ``num_buckets`` is a power of 2). So, the most important bits of the hash function are the least significant bits.